ocaml/xapi/livepatch.ml

```diff
@@ -0,0 +1,343 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "livepatch" end)
+
+open D
+
+exception Invalid_component
+
+type component = Xen | Kernel
+
+let component_of_string = function
+  | "Xen" | "xen" ->
+      Xen
+  | "Kernel" | "kernel" ->
+      Kernel
+  | _ ->
+      raise Invalid_component
+
+let pkg_name_of_component = function
+  | Xen ->
+      "xen-hypervisor"
+  | Kernel ->
+      "kernel"
+
+let string_of_component = function Xen -> "xen" | Kernel -> "kernel"
+
+type t = {
+    component: component
+  ; base_build_id: string
+  ; base_version: string
+  ; base_release: string
+  ; to_version: string option
+  ; to_release: string option
+}
+
+let to_json lp =
+  let option_to_json o =
+    Option.(value (map (fun x -> `String x) o) ~default:`Null)
+  in
+  `Assoc
+    [
+      ("component", `String (string_of_component lp.component))
+    ; ("base_build_id", `String lp.base_build_id)
+    ; ("base_version", `String lp.base_version)
+    ; ("base_release", `String lp.base_release)
+    ; ("to_version", option_to_json lp.to_version)
+    ; ("to_release", option_to_json lp.to_release)
+    ]
+
+let of_json js =
+  try
+    let open Yojson.Basic.Util in
+    {
+      component= member "component" js |> to_string |> component_of_string
+    ; base_build_id= member "base_build_id" js |> to_string
+    ; base_version= member "base_version" js |> to_string
+    ; base_release= member "base_release" js |> to_string
+    ; to_version= member "to_version" js |> to_string_option
+    ; to_release= member "to_release" js |> to_string_option
+    }
+  with e ->
+    let msg = "Can't construct a livepatch from json" in
```

Comment:
This should either include a path to the JSON or the JSON itself to facilitate debugging.

Reply:
This is now logged below.